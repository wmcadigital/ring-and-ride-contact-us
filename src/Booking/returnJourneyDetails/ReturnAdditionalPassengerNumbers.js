import { useFormState, useForm } from "react-final-form";

import FormSection from "../../common/FormSection";
import Question from "../../common/Question";
import ProgressIndicator from "../../common/ProgressIndicator";
import NumberInput from "../../common/NumberInput";
import getSectionPositionInfo from "../getSectionPosition";
import getAboutReturnJourneySectionName from "./getAboutReturnJourneySectionName";
import {
  required,
  numbersOnly,
  numberGreaterThanZero,
  composeValidators,
} from "../../common/validation";

const ReturnAdditionalPassengerNumbers = () => {
  const stateApi = useFormState();
  const formApi = useForm();

  const formValues = stateApi.values;
  const bookingParty = formValues["bookingParty"];
  const passengerReturnNumbers = formValues["additionalReturnPassengerNumbers"];

  let question = "";
  if (bookingParty === "mySelf") {
    question = "How many people are joining you on the return journey?";
  } else {
    question = `How many people are joining ${formValues["firstName"]} on the return journey?`;
  }

  const pronoun = formValues["bookingParty"] === "mySelf" ? "you" : "them";

  const error = stateApi.submitFailed
    ? stateApi.errors?.["additionalReturnPassengerNumbers"]
    : null;

  const decreaseNumberCallback = () =>
    formApi.mutators.setFormAttribute(
      "additionalReturnPassengerNumbers",
      passengerReturnNumbers - 1
    );
  const increaseNumberCallback = () =>
    formApi.mutators.setFormAttribute(
      "additionalReturnPassengerNumbers",
      passengerReturnNumbers + 1
    );

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition={getSectionPositionInfo("3", stateApi)}
        sectionName={getAboutReturnJourneySectionName(stateApi)}
      />
      <Question text={question} />
      <NumberInput
        fieldName="additionalReturnPassengerNumbers"
        label={`Number of people joining ${pronoun} on the journey`}
        defaultValue={1}
        validation={composeValidators(
          required,
          numbersOnly,
          numberGreaterThanZero
        )}
        decreaseNumberCallback={decreaseNumberCallback}
        increaseNumberCallback={increaseNumberCallback}
        error={error}
      />
    </FormSection>
  );
};

export default ReturnAdditionalPassengerNumbers;
