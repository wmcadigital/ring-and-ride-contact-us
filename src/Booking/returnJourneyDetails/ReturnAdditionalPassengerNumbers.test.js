import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import ReturnAdditionalPassengerNumbers from "./ReturnAdditionalPassengerNumbers";

jest.mock("react-final-form");

describe("ReturnAdditionalPassengerNumbers - snapshots", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<ReturnAdditionalPassengerNumbers />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "behalfSomeone",
        firstName: "John",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<ReturnAdditionalPassengerNumbers />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
        firstName: "John",
        additionalReturnPassengerNumbers: -1,
      },
      submitFailed: true,
      errors: { additionalReturnPassengerNumbers: "Invalid" },
    }));
    const renderer = createRenderer();

    renderer.render(<ReturnAdditionalPassengerNumbers />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
