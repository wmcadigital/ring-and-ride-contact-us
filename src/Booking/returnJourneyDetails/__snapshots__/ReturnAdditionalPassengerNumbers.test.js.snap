// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReturnAdditionalPassengerNumbers - snapshots individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan their return journey"
    sectionPosition="Section 3 of 4"
  />
  <Question
    text="How many people are joining John on the return journey?"
  />
  <NumberInput
    decreaseNumberCallback={[Function]}
    defaultValue={1}
    error={null}
    fieldName="additionalReturnPassengerNumbers"
    increaseNumberCallback={[Function]}
    label="Number of people joining them on the journey"
    validation={[Function]}
  />
</FormSection>
`;

exports[`ReturnAdditionalPassengerNumbers - snapshots individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan your return journey"
    sectionPosition="Section 3 of 4"
  />
  <Question
    text="How many people are joining you on the return journey?"
  />
  <NumberInput
    decreaseNumberCallback={[Function]}
    defaultValue={1}
    error={null}
    fieldName="additionalReturnPassengerNumbers"
    increaseNumberCallback={[Function]}
    label="Number of people joining you on the journey"
    validation={[Function]}
  />
</FormSection>
`;

exports[`ReturnAdditionalPassengerNumbers - snapshots renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan your return journey"
    sectionPosition="Section 3 of 4"
  />
  <Question
    text="How many people are joining you on the return journey?"
  />
  <NumberInput
    decreaseNumberCallback={[Function]}
    defaultValue={1}
    error="Invalid"
    fieldName="additionalReturnPassengerNumbers"
    increaseNumberCallback={[Function]}
    label="Number of people joining you on the journey"
    validation={[Function]}
  />
</FormSection>
`;
